### PR TITLE
Support the full MCTP protocol

### PR DESCRIPTION
**Pull request by alistair23**

### CODE REVIEW

**Comment by twilfredo** [comment]
Nice, did you have a go at testing this?

**Comment by alistair23** [comment]
It runs for me, but eventually I get an error sending the data. I *think* that's an I2C issue and not related to this change

**Comment by twilfredo** [comment]
> It runs for me, but eventually I get an error sending the data. I _think_ that's an I2C issue and not related to this change

Okay, I'll run some tests soon... then merge if all good. 

Should we add  some scripted unit tests for hardware testing, basically something to invoke spdm-utils and do assertions on the outputs?

**Comment by alistair23** [comment]
> Should we add some scripted unit tests for hardware testing, basically something to invoke spdm-utils and do assertions on the outputs?

Yeah, that would be handy!

File: src/usb_i2c.rs [hunk 1]
```diff
@@ -157,17 +181,34 @@ unsafe extern "C" fn usb_i2c_receive_message(
...
+
+    // Ammed our address to create a valid MCTP packet
+    spdm_msg_buf.copy_within(0..message_len, 1);
+    spdm_msg_buf[0] = SOURCE_ID << 1;
+    message_len = message_len - 1;
+
+    let mctp_len = spdm_msg_buf[2] as usize + 4;
+    assert!(mctp_len <= message_len);
+
+    let (_msg_type, payload) = ctx.decode_packet(&spdm_msg_buf[0..mctp_len]).unwrap();
```
Comment (twilfredo) [inline src/usb_i2c.rs:201]: Running into a panic in `libmctp` when I issue a request from `spdm-utils`

```
                          Particle Boron: Initialization complete. Entering main loop
spdm-sample: app start
spdm-sample: setup_transport_layer [ok]
spdm-sample: setup_capabilities [ok]
spdm-sample: starting response_loop...
panicked at 'range end index 8 out of range for slice of length 6', /home/twilfred/.cargo/git/checkouts/libmctp-e88d4618e42006d6/f9b4473/src/smbus.rs:224:42
```

It's called by the call to `decode_packet` in the case where the slice passed in is has < 8 indexes. Call path in libmctp -> decode_packet() -> get_smbus_headers(), then panic: `base_header_buf.copy_from_slice(&packet[4..8]);`

**Review by twilfredo — CHANGES_REQUESTED** [review]

File: tock-responder/src/mctp.rs [hunk 2]
```diff
@@ -173,32 +212,56 @@ unsafe extern "C" fn tock_receive_message(
...
-            r.0,
-            &recv_buf[0..r.0]
+            message_len,
+            &recv_buf[0..message_len]
         )
         .unwrap();
     }
 
-    *message_size = r.0;
+    let (_msg_type, payload) = ctx.decode_packet(&recv_buf[0..message_len]).unwrap();
```
Comment (twilfredo) [inline tock-responder/src/mctp.rs:251]: Running into a panic in `libmctp` when I issue a request from `spdm-utils`

```
                          Particle Boron: Initialization complete. Entering main loop
spdm-sample: app start
spdm-sample: setup_transport_layer [ok]
spdm-sample: setup_capabilities [ok]
spdm-sample: starting response_loop...
panicked at 'range end index 8 out of range for slice of length 6', /home/twilfred/.cargo/git/checkouts/libmctp-e88d4618e42006d6/f9b4473/src/smbus.rs:224:42
```

It's called by the call to `decode_packet` in the case where the slice passed in has < 8 indexes (`&recv_buf[0..message_len]`). Call path in libmctp -> decode_packet() -> get_smbus_headers(), then panic `L224:smbus.rs`: `base_header_buf.copy_from_slice(&packet[4..8]);`

**Review by twilfredo — CHANGES_REQUESTED** [review]

File: tock-responder/src/mctp.rs (hunk 2 above)
Comment (twilfredo) [reply tock-responder/src/mctp.rs:251]: Updated: with `spdm-debug` mode outputs

Comment (alistair23) [reply tock-responder/src/mctp.rs:251]: `[44, 5, 10, 84, 0, 0]` isn't a valid MCTP packet, it looks like just a plain SPDM payload. Something odd is going on here

File: src/usb_i2c.rs [hunk 3]
```diff
@@ -157,17 +181,34 @@ unsafe extern "C" fn usb_i2c_receive_message(
...
-    debug!("spdm_msg_data: {:x?}", *message_size);
+    port.read_exact(&mut spdm_msg_buf[0..message_len]).unwrap();
+
+    // Ammed our address to create a valid MCTP packet
+    spdm_msg_buf.copy_within(0..message_len, 1);
+    spdm_msg_buf[0] = SOURCE_ID << 1;
+    message_len = message_len - 1;
+
+    let mctp_len = spdm_msg_buf[2] as usize + 4;
+    assert!(mctp_len <= message_len);
```
Comment (twilfredo) [inline src/usb_i2c.rs:199]: this assertion fails, how do we know what `spdm_msg_buf[2]` is?

File: src/usb_i2c.rs [hunk 4]
```diff
@@ -157,17 +181,34 @@ unsafe extern "C" fn usb_i2c_receive_message(
...
+
+    let (_msg_type, payload) = ctx.decode_packet(&spdm_msg_buf[0..mctp_len]).unwrap();
 
+    let len = payload.len() + 1;
+    spdm_msg_buf.copy_within(9..(9 + len), 1);
```
Comment (twilfredo) [inline src/usb_i2c.rs:204]: what is happening here? this is hard to follow?

File: tock-responder/src/mctp.rs [hunk 5]
```diff
@@ -173,32 +212,56 @@ unsafe extern "C" fn tock_receive_message(
...
+
+    let _ = MCTPCONTEXT.set(ctx);
+
+    let len = payload.len() + 1;
+    recv_buf.copy_within(9..(9 + len), 1);
```
Comment (twilfredo) [inline tock-responder/src/mctp.rs:256]: might be better make these into consts so we know what the offsets are doing (same as above).

File: tock-responder/src/mctp.rs (hunk 2 above)
Comment (twilfredo) [reply tock-responder/src/mctp.rs:251]: my bad, I forgot to rebuild `spdm-utils` :P

File: tock-responder/src/mctp.rs (hunk 5 above)
Comment (twilfredo) [reply tock-responder/src/mctp.rs:256]: or some comments

File: src/usb_i2c.rs (hunk 3 above)
Comment (alistair23) [reply src/usb_i2c.rs:199]: It's the byte count of an MCTP packet. What values are you seeing in the failure?

Comment (twilfredo) [reply src/usb_i2c.rs:199]: I put a debug to print out the two values:
```
[2024-01-22T23:32:18Z DEBUG spdm_utils::usb_i2c] mctp_len: 18  message_len: 16
thread 'main' panicked at 'assertion failed: mctp_len <= message_len', src/usb_i2c.rs:200:5
```

**Review by twilfredo — APPROVED** [review]
